[agent] Check for hidepid=2 on proc mount and report in logs

xornivore opened the pull request:
### What does this PR do?

With `hidepid=2` on proc mount the Agent will not be able to see other processes. This adds a warning in the logs so that we can detect these configurations. In the future we can expose this in the agent status as well.

More info on [hidepid](https://man7.org/linux/man-pages/man5/proc.5.html).

### Motivation

Detecting `hidepid=2` configuration.

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan

Harden the host with `hidepid=2,gid=..` and test that the message is shown in the logs.

Commit:
[agent] Check for hidepid=2 on proc mount and report as misconfig in logs

L3n41c commented on cmd/agent/common/misconfig/global.go:
```diff
@@ -0,0 +1,22 @@
+func ToLog() {
+	for name, check := range checks {
+		if err := check(); err != nil {
+			log.Warnf("misconfig: %s: %v", name, err)
```
```suggestion
			log.Warnf("misconfig: %s: %w", name, err)
```

xornivore replied:
I suspect it is a bit less subtle here - we aren't wrapping an error, just logging it as opposed to `errors.Wrapf`

L3n41c commented on cmd/agent/common/misconfig/mounts.go:
```diff
@@ -0,0 +1,65 @@
+	if err != nil {
+		return fmt.Errorf("failed to get process groups: %v", err)
+	}
+	return checkProcMountHidePid("/host/proc/1/mounts", groups)
```
```suggestion
	return checkProcMountHidePid(filepath.Join(config.Datadog.GetString("container_proc_root"), "1/mounts"), groups)
```

I’m wondering if we should mimic [this logic (found in `integrations-core`)](https://github.com/DataDog/integrations-core/blob/22dd4c4015cdd30aa7a778076683d1b01b90b862/network/datadog_checks/network/network.py#L489-L493) where the `/1/` part is added only in case we have an evidence that the agent is containerised.

vboulineau reviewed (approved):
LGTM

L3n41c commented on cmd/agent/common/misconfig/mounts.go:
```diff
@@ -0,0 +1,65 @@
+func checkProcMountHidePid(path string, groups []int) error {
+	file, err := os.Open(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
```
Why do we want to be silent when `/host/proc/1/mounts` doesn’t exist ?
If it doesn’t exist, there must be a misconfiguration that should be raised to the user.
I suppose it’s for not raising an error when the agent isn’t running inside a container and `/host/proc` doesn’t exist.
I think that even when the agent isn’t running inside a container, it makes sense to check for the `hidepid` option on `/proc` (instead of `/host/proc`).
So, I would be in favour of detecting whether we should check `/proc/mounts` or `/host/proc/1/mounts` depending on the context and to always raise an error.

vboulineau replied:
The issue is more with environments where we don't have any meaningful `/proc`, like Fargate.

Though I think `config.Datadog.GetString("container_proc_root")` returns `/proc` in this case, which should make it work there too.

L3n41c commented on cmd/agent/common/misconfig/mounts.go:
```diff
@@ -0,0 +1,65 @@
+	for scanner.Scan() {
+		fields := strings.Fields(scanner.Text())
+		if len(fields) != 6 ||
+			fields[0] != "proc" || fields[1] != "/proc" {
```
We shouldn’t check the first column (`fields[0] != "proc"`) because the first column is the device to mount and, as `proc` is a pseudo-filesystem, we can put anything as device.

For ex.:
```
root@agent-dev-lenaic:~# ls /tmp/toto/
1    138  149  159  20    2505  3    34    384   4     464   5795  5961  6103  652  795        cmdline    execdomains  kallsyms     loadavg  net           slabinfo       timer_list
10   139  15   16   21    259   30   35    385   40    465   5822  5970  628   663  814        consoles   fb           kcore        locks    pagetypeinfo  softirqs       tty
[…]
root@agent-dev-lenaic:~# grep toto /proc/mounts
anything /tmp/toto proc rw,relatime 0 0
```

On the other hand, we should check the third column (`fields[2]`), because this is the filesystem and here, we want to check the `proc` filesystem.

L3n41c commented on cmd/agent/common/misconfig/mounts.go:
```diff
@@ -0,0 +1,65 @@
+			continue
+		}
+		mountOpts := strings.Split(fields[3], ",")
+		mountOptsLookup := map[string]bool{}
```
```suggestion
		mountOptsLookup := map[string]struct{}
```

For sets, I use `map[string]struct{}` in Go. It saves few extra memory bytes since there is no `bool` value to store. ;-)

L3n41c commented on cmd/agent/common/misconfig/mounts.go:
```diff
@@ -0,0 +1,65 @@
+		mountOpts := strings.Split(fields[3], ",")
+		mountOptsLookup := map[string]bool{}
+		for _, opt := range mountOpts {
+			mountOptsLookup[opt] = true
```
```suggestion
			mountOptsLookup[opt] = struct{}{}
```

L3n41c commented on cmd/agent/common/misconfig/mounts.go:
```diff
@@ -0,0 +1,65 @@
+			mountOptsLookup[opt] = true
+		}
+
+		if ok, _ := mountOptsLookup["hidepid=2"]; !ok {
```
```suggestion
		if _, ok := mountOptsLookup["hidepid=2"]; !ok {
```

Commit:
Address review feedback

L3n41c reviewed (approved):
Awesome 👍🏻

olivielpeau reviewed (approved):
LGTM, but do we expect cases where a user would know about this, but would not want to get warned about it? (in non-containerized environments I guess most of the Agent would still work with `hidepid=2`). If so, having an opt-in config option to disable a set of checks would be nice.

xornivore commented:
Thank you @olivielpeau , do you think it might be better if we check that process agent or other features (e.g. continuous compliance) are enabled before checking here? That would guard by enabled features as opposed to introducing another config setting (arguably a bit harder to understand from a user's standpoint).

olivielpeau commented:
@xornivore yes makes sense to me if you have a good sense of which agent features would be affected by `hidepid=2` 👍

xornivore commented:
> @xornivore yes makes sense to me if you have a good sense of which agent features would be affected by `hidepid=2` 👍

@olivielpeau, unfortunately as @vboulineau pointed out to me it is will not always be possible, e.g. for basic cases like Docker check on a host. Going to keep as is for now but later we could add `misconfig.skip_checks` or a similarly named setting.

olivielpeau commented:
ok, thanks for checking! Since the warning is at most logged once on agent startup, I'm OK with merging as-is without a config option disabling it 👍

Commit:
Adjust the misconfig logic to account for root with gid= setting

Commit:
Update hidepid logic to reflect behavior more accurately

- If we are running with root (euid=0) and cap_sys_ptrace - no need to check for hidepid/gid.
- Otherwise same logic applies for root and non root
- Correct build flag to linux only.

Commit:
More meaningful logs + fix for egid not being included in the list returned by `getgroups`.

Commit:
Add `// nolint: deadcode, unused` to fix macos test failures
